Python/optimizer.c

```diff
@@ -732,6 +754,11 @@ translate_bytecode_to_trace(
         instr++;
         // Add cache size for opcode
         instr += _PyOpcode_Caches[_PyOpcode_Deopt[opcode]];
+        /* Stop if we have reached the start */
+        if (instr == initial_instr) {
+            RESERVE(1);
+            ADD_TO_TRACE(_JUMP_TO_TOP, 0, 0, 0);
+        }
     }  // End for (;;)
 
 done:
```

Comment:
Shouldn't we check here that `progress_needed` is false and `return 0;` if it's still true? Maybe it's covered implicitly from the `trace_length` check, but that's not clear if so (and probably not future-proof).